builder-pattern-macro/src/struct_impl.rs

```diff
@@ -15,34 +18,44 @@ impl<'a> ToTokens for StructImpl<'a> {
     fn to_tokens(&self, tokens: &mut TokenStream) {
         let ident = &self.input.ident;
         let vis = &self.input.vis;
-        let where_clause = &self.input.generics.where_clause;
         let builder_name = self.input.builder_name();
 
         let lifetimes = self.input.lifetimes();
-        let impl_tokens = self.input.tokenize_impl();
         let empty_generics = self.empty_generics();
-        let ty_tokens = self.input.tokenize_types();
+        let defaulted_generics = self.input.defaulted_generics();
+
+        let with_prmdef = |ident: &Ident| self.input.with_param_default(&defaulted_generics, ident);
+        let replace_defaults =
+            |stream: TokenStream| replace_type_params_in(stream, &defaulted_generics, &with_prmdef);
+
+        let impl_tokens = self.input.tokenize_impl(&defaulted_generics);
+
+        let where_clause = &self.input.generics.where_clause;
+        let where_tokens =
+            replace_type_params_in(quote! { #where_clause }, &defaulted_generics, &with_prmdef);
+
+        let ty_tokens = replace_defaults(self.input.tokenize_types(&[], false));
 
         let fn_lifetime = self.input.fn_lifetime();
 
         let builder_init_args = self.builder_init_args();
         let docs = self.documents();
 
         tokens.extend(quote! {
-            impl <#impl_tokens> #ident <#(#lifetimes,)* #ty_tokens> #where_clause {
+            impl <#impl_tokens> #ident <#(#lifetimes,)* #ty_tokens> #where_tokens {
```

Comment:
```rust
struct X<A, B = f64> {
    a: A,
    b: B
}
```
It will generate the following code for the above struct:

```rust
impl<A> X<A, f64> {
    fn new<'fn_lifetime>() -> XBuilder<'fn_lifetime, A, f64, (), (), (), (), ()> {}
}

X::<i32>::new(); // ok
X::new().a(3); // ok
X::new().a(3).b(5); // expected: f64, given: i32
X::<i32, i32>::new(); // No associated function `new`
```

For the as-is way, it will generate the following code:

```rust
impl<A, B> X<A, B> {
    fn new<'fn_lifetime>() -> XBuilder<'fn_lifetime, A, B, (), (), (), (), ()> {}
}

X::<i32>::new(); // ok
X::new().a(3); // Cannot infer B
X::new().a(3).b(5); // ok
X::<i32, i32>::new(); // ok
```

I think the previous way can handle more cases. Is this change required for implementing the `infer` attribute?

Reply:
The ability to choose which ones are defaulted in the `new` function would be quite nice. Here's a comparison of two methods of choosing.

| Method | Breaking change? | Behaviour without customization | what the attribute does |
| --- | --- | --- | --- |
| #[new_infer(B)] struct X... | yes | like std's HashMap, Vec, etc | allows inference of B in `new`, means `#[default(...)] b: B` stops working |
| #[new_default(B)] struct X... | no | like existing builder-pattern, not like std | substitutes B=f64 in `new`, allows `#[default(...)] b: B` to work |

`#[new_infer]` also matches up with `#[infer]`, learn it once and understand both. Whereas there is no way to name the `#[new_default]` attribute in a way that doesn't require reading docs eight times, because there are already `#[default]` attributes that do completely different stuff.